### ISSUE TITLE
tif分类图中的错分很多

### ISSUE THREAD

**Issue by immortal11111** [virylon/CNN_Pavia#2]
![image](https://user-images.githubusercontent.com/58716529/70492402-47388f00-1b3f-11ea-82fd-973089a54a66.png)

您好，非常感谢您分享的代码！我用CPU训练时间为23h，测试和显示分类图的时间为16h，train set Accuracy:99.98%，test set Accuracy:99.66%，分类精度很高。
但是我在ArcGIS中看分类图时，却发现分类图有很多错分，这是为什么呢？

**Comment by inspire-lts** [comment]
@immortal11111  请问下我为什么训练显示出来是黑白的勒，难道次数少了吗？刚入门不咋懂

**Comment by virylon** [comment]
这个精度并不是很高，跟模型还有训练的参数有关系，由于将背景忽略还有逐像元分类的方式，椒盐现象严重，需要分类后进行聚类处理。如果都是黑白的可能是由于是灰度值显示方式，可以用其他颜色模型显示

**Comment by inspire-lts** [comment]
![image](https://user-images.githubusercontent.com/33800259/70721658-f38d9780-1d30-11ea-89f5-cd17ecb4d87f.png)
十分感谢 这是我显示出来的图片   请问我想显示出不同的类显示出不同的颜色注释掉您show函数里面的就可以了？

**Comment by immortal11111** [comment]
> @immortal11111 请问下我为什么训练显示出来是黑白的勒，难道次数少了吗？刚入门不咋懂

用原始代码训练出来是黑白的tif，有两种方式将其变成彩色图片：
1.用arcgis唯一值显示，这还可以去掉那些没有标注的像元；
2.`from scipy import misc
import spectral
tif32=misc.imread(r"F:\CNN_Pavia-master\PU\predicted\12-08_13-09-38.tif")
gl = spectral.imshow(classes = tif32.astype(int),figsize=(16,16))
`然后在cmd上写'ipython --pylab'，再写%run C:\Users\XXX\Desktop\show.py就行了

**Comment by immortal11111** [comment]
> 这个精度并不是很高，跟模型还有训练的参数有关系，由于将背景忽略还有逐像元分类的方式，椒盐现象严重，需要分类后进行聚类处理。如果都是黑白的可能是由于是灰度值显示方式，可以用其他颜色模型显示
> […](#)
> Sent from my iPhone
> On Dec 12, 2019, at 21:32, inspire-lts ***@***.***> wrote: ﻿ @immortal11111 请问下我为什么训练显示出来是黑白的勒，难道次数少了吗？刚入门不咋懂 — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub, or unsubscribe.

Thanks！我还是有些问题😛😛，最终显示的训练集和测试集的分类精度达到了99%以上，分类图理应与其相对应呀。现在是不是大多数算法都是逐项元分类呀，面向对象的分类很少是吗，忽略背景该如何理解呢

**Comment by inspire-lts** [comment]
@immortal11111  谢谢啦！

**Comment by DAYdreaming110** [comment]
作者大大能不能教我运行这个代码呢？  我环境也配置了 数据也安排上了 但是为什么还是运行不了呢？

**Comment by virylon** [comment]
> 作者大大能不能教我运行这个代码呢？ 我环境也配置了 数据也安排上了 但是为什么还是运行不了呢？

可以把问题分享出来看看呢？

**Comment by DAYdreaming110** [comment]
![image](https://user-images.githubusercontent.com/101042746/162215724-11ae1928-5c47-4a47-8d64-d50d2d2d2392.png)

**Comment by DAYdreaming110** [comment]
作者大大方不方便留个联系方式   qq 1251072503

**Comment by huanyu233** [comment]
训练模型时分类地物数量是9，但是说IndexError: index 9 is out of bounds for axis l with size 9，我看patch切片里有0~8个测试模型，0~8个训练模型，在train.py将classes_name = [str(c) for c in range(9)] 变为classes_name = [str(c) for c in range(16)] ,epoch5次之后精度很低，为什么呀 求大大解惑  qq1047636161

**Comment by virylon** [comment]
> 训练模型时分类地物数量是9，但是说IndexError: index 9 is out of bounds for axis l with size 9，我看patch切片里有0~8个测试模型，0~8个训练模型，在train.py将classes_name = [str(c) for c in range(9)] 变为classes_name = [str(c) for c in range(16)] ,epoch5次之后精度很低，为什么呀 求大大解惑 qq1047636161

**Comment by virylon** [comment]
> 

你用的是自己的数据吗？要分类的地物有16种？

**Comment by virylon** [comment]
> 训练模型时分类地物数量是9，但是说IndexError: index 9 is out of bounds for axis l with size 9，我看patch切片里有0~8个测试模型，0~8个训练模型，在train.py将classes_name = [str(c) for c in range(9)] 变为classes_name = [str(c) for c in range(16)] ,epoch5次之后精度很低，为什么呀 求大大解惑 qq1047636161

你用的是自己制作的数据吗？要分类的地物有16种？

**Comment by huanyu233** [comment]
不是的，我觉得很奇怪，patch切片处理之后数据集是0~8个测试集，0~8个训练集，按道理来说测试集是9个，classes_name = [str(c) for c in range(9)] 第一次迭代之后就显示IndexError: index 9 is out of bounds for axis l with size 9，从10，11，12，一个一个是都失败了，只有16迭代成功了，但是精确率只有百分之三点几，

**Comment by huanyu233** [comment]
我的问题解决了optimizer.step()` 和lr_scheduler.step()的原因，因为我的pytorch版本属于PyTorch 1.1.0以后，它弹出的警告UserWarning: 检测到在`optimizer.step()`之前调用了`lr_scheduler.step()`。在PyTorch 1.1.0及以后的版本中，您应该以相反的顺序调用它们。`optimizer.step()`在`lr_scheduler.step()`之前。 如果不这样做，将导致PyTorch跳过学习率计划的第一个值。更多细节请参见https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  warnings.warning("检测到在`optimizer.step()`之前调用`lr_scheduler.step()`。"我把源代码的更新权值和更新学习率调换了，警告没了，但是学习率很低，而且出现以上错误。其实可以忽略这个警告，用源代码即可

**Comment by virylon** [comment]
> 不是的，我觉得很奇怪，patch切片处理之后数据集是0~8个测试集，0~8个训练集，按道理来说测试集是9个，classes_name = [str(c) for c in range(9)] 第一次迭代之后就显示IndexError: index 9 is out of bounds for axis l with size 9，从10，11，12，一个一个是都失败了，只有16迭代成功了，但是精确率只有百分之三点几，

QQ 3469166620
